From fe89a6a1e6e45c10e98627b8e879ca8c3256db97 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 26 Jun 2014 15:33:35 -0400
Subject: [fs] nfs: pass string length to pr_notice message about readdir loops

Message-id: <1403796815-28580-30-git-send-email-steved@redhat.com>
Patchwork-id: 84869
O-Subject: [RHEL7.1 PATCH 29/29] nfs: pass string length to pr_notice message about readdir loops
Bugzilla: 1111170
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Jeff Layton <jlayton@redhat.com>

There is no guarantee that the strings in the nfs_cache_array will be
NULL-terminated. In the event that we end up hitting a readdir loop, we
need to ensure that we pass the warning message the length of the
string.

Reported-by: Lachlan McIlroy <lmcilroy@redhat.com>
Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upsteam Id: 9581a4ae75517099bc87e1c43d1a8f35b55741b9
Bugzilla: 1111170
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/dir.c b/fs/nfs/dir.c
index 7eaaf7d..330ec1a 100644
--- a/fs/nfs/dir.c
+++ b/fs/nfs/dir.c
@@ -312,10 +312,9 @@ int nfs_readdir_search_for_cookie(struct nfs_cache_array *array, nfs_readdir_des
      if (printk_ratelimit()) {
       pr_notice("NFS: directory %pD2 contains a readdir loop."
         "Please contact your server vendor.  "
-        "The file: %s has duplicate cookie %llu\n",
-        desc->file,
-        array->array[i].string.name,
-        *desc->dir_cookie);
+        "The file: %.*s has duplicate cookie %llu\n",
+        desc->file, array->array[i].string.len,
+        array->array[i].string.name, *desc->dir_cookie);
      }
      status = -ELOOP;
      goto out;
-- 
1.7.1